modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceProcessorAdapter.java

```diff
@@ -0,0 +1,155 @@
+     */
+    public abstract IgniteInternalFuture<?> deployMultiple(ClusterGroup prj, String name, Service srvc, int totalCnt,
+        int maxPerNodeCnt);
+
+    /**
+     * @param name Service name.
+     * @param srvc Service.
+     * @param cacheName Cache name.
+     * @param affKey Affinity key.
+     * @return Future.
+     */
+    public abstract IgniteInternalFuture<?> deployKeyAffinitySingleton(String name, Service srvc, String cacheName,
+        Object affKey);
+
+    /**
+     * @param prj Grid projection.
+     * @param cfgs Service configurations.
+     * @return Future for deployment.
+     */
+    public abstract IgniteInternalFuture<?> deployAll(ClusterGroup prj, Collection<ServiceConfiguration> cfgs);
+
+    /**
+     * @param name Service name.
+     * @return Future.
+     */
+    public abstract IgniteInternalFuture<?> cancel(String name);
+
+    /**
+     * @return Future.
+     */
+    public abstract IgniteInternalFuture<?> cancelAll();
+
+    /**
+     * @param servicesNames Name of services to deploy.
+     * @return Future.
+     */
+    public abstract IgniteInternalFuture<?> cancelAll(Collection<String> servicesNames);
+
+    /**
+     * @return Collection of service descriptors.
+     */
+    public abstract Collection<ServiceDescriptor> serviceDescriptors();
+
+    /**
+     * @param name Service name.
+     * @param <T> Service type.
+     * @return Service by specified service name.
+     */
+    public abstract <T> T service(String name);
+
+    /**
+     * @param prj Grid projection.
+     * @param name Service name.
+     * @param srvcCls Service class.
+     * @param sticky Whether multi-node request should be done.
+     * @param timeout If greater than 0 limits service acquire time. Cannot be negative.
+     * @param <T> Service interface type.
+     * @return The proxy of a service by its name and class.
+     * @throws IgniteException If failed to create proxy.
+     */
+    public abstract <T> T serviceProxy(ClusterGroup prj, String name, Class<? super T> srvcCls, boolean sticky,
+        long timeout) throws IgniteException;
+
+    /**
+     * @param name Service name.
+     * @param <T> Service type.
+     * @return Services by specified service name.
+     */
+    public abstract <T> Collection<T> services(String name);
+
+    /**
+     * @param name Service name.
+     * @return Service by specified service name.
+     */
+    public abstract ServiceContextImpl serviceContext(String name);
+
+    /**
+     * @param name Service name.
+     * @param timeout If greater than 0 limits task execution time. Cannot be negative.
+     * @return Service topology.
+     * @throws IgniteCheckedException On error.
+     */
+    public abstract Map<UUID, Integer> serviceTopology(String name, long timeout) throws IgniteCheckedException;
+
+    /**
+     * @return {@code true} if event driven service processor enabled, otherwise {@code false};
+     */
+    public boolean eventDrivenServiceProcessorEnabled() {
```

Comment:
So, this method is supposed to be removed together with the old `GridServiceProcessor`, right? So it's kind of deprecated and doomed?

I think it would be better to remove the method and all related checks. If you need to do something only in the new processor (e.g. `localNodeJoin`), just and a no-op implementation to the old one. If you need to do something only in the old one, just check with `srvcProc instanceof GridServiceProcessor` - the ugliness of the check will match the deprecated status of the old processor.

Reply:
This method has been introduced to be called in heavy loaded place to avoid performance issues with `instance of` operator. But these places have been reworked and the remained places are not hot, so it can be replaced `instance of`.
I removed this method.
Thanks!

Reply:
Can you elaborate on the "performance issues with `instanceof`"? Never heard of it being a thing. I'd say that JIT should optimize that to `true`/`false` quite easily, especially since you're going to only have instances of one of the service processor implementations in the JVM.